src/riak_kv_replrtq_peer.erl

```diff
@@ -0,0 +1,240 @@
+%% software distributed under the License is distributed on an
+%% "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+%% KIND, either express or implied.  See the License for the
+%% specific language governing permissions and limitations
+%% under the License.
+%%
+%% -------------------------------------------------------------------
+
+%% @doc Peer discovery for real-time replication
+
+-module(riak_kv_replrtq_peer).
+-ifdef(TEST).
+-include_lib("eunit/include/eunit.hrl").
+-endif.
+
+-behaviour(gen_server).
+
+-export([init/1,
+        handle_call/3,
+        handle_cast/2,
+        handle_info/2,
+        terminate/2,
+        code_change/3]).
+
+-export([start_link/0,
+            update_discovery/1,
+            prompt_discovery/1,
+            update_workers/2]).
+
+-type discovery_peer() ::
+    {riak_kv_replrtq_snk:queue_name(), [riak_kv_replrtq_snk:peer_info()]}.
+
+-define(DISCOVERY_TIMEOUT_SECONDS, 60).
+-define(AUTO_DISCOVERY_MAXIMUM_SECONDS, 900).
+-define(AUTO_DISCOVERY_MINIMUM_SECONDS, 60).
+
+-record(state, {discovery_peers = [] :: list(discovery_peer())}).
+
+%%%============================================================================
+%%% API
+%%%============================================================================
+
+-spec start_link() -> {ok, pid()}.
+start_link() ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
+
+%% @doc
+%% Prompt for the discovery of peers
+-spec update_discovery(riak_kv_replrtq_snk:queue_name()) -> boolean().
+update_discovery(QueueName) ->
+    gen_server:call(?MODULE, {update_discovery, QueueName}, 60 * 1000).
+
+-spec update_workers(pos_integer(), pos_integer()) -> boolean().
+update_workers(WorkerCount, PerPeerLimit) ->
+    gen_server:call(?MODULE,
+        {update_workers, WorkerCount, PerPeerLimit},
+        60 * 1000).
+
+-spec prompt_discovery(discovery_peer()) -> ok.
+prompt_discovery({QueueName, PeerInfo}) ->
+    gen_server:cast(?MODULE, {prompt_discovery, {QueueName, PeerInfo}}).
+
+
+%%%============================================================================
+%%% gen_server callbacks
+%%%============================================================================
+
+init([]) ->
+    case app_helper:get_env(riak_kv, replrtq_peer_discovery, false) of
+        true -> 
+            SinkPeers = app_helper:get_env(riak_kv, replrtq_sinkpeers, ""),
+            DefaultQueue = app_helper:get_env(riak_kv, replrtq_sinkqueue),
+            SnkQueuePeerInfo =
+                riak_kv_replrtq_snk:tokenise_peers(DefaultQueue, SinkPeers),
+
+            MinDelay = 
+                app_helper:get_env(riak_kv,
+                    replrtq_prompt_min_seconds,
+                    ?AUTO_DISCOVERY_MINIMUM_SECONDS),
+
+            lists:foreach(
+                fun({QueueName, _PeerInfo}) -> 
+                    erlang:send_after(MinDelay * 1000,
+                        self(),
+                        {prompt_discovery, QueueName})
```

Comment:
It takes a few moments to realize that there are 2 interfaces for `prompt discovery` where this one goes via `handle_info` and has only one argument that does a lookup for the PeerInfo at the time it executes, whereas the cast version gets the PeerInfo in the interface.


Probably one should comment this difference more clearly.